### PR TITLE
Basic verification for responses

### PR DESCRIPTION
**Pull request by sztomi**
This PR adds a new crate called `integration_tests` which has a build script that generates tests from all response variants. The tests have two forms:

For variants that don't hold data:

```rust
  #[test]
  fn validate_terminatethreads_response() {
    let resp = Response {
      request_seq: 1,
      success: true,
      message: None,
      body: Some(ResponseBody::TerminateThreads),
    };
    let value = resp_to_value(&resp);
    assert!(SCHEMA.is_valid(&value));
  }
```

For variants that hold data:

```rust
  #[test]
  fn validate_threads_response() {
    let resp = Response {
      request_seq: 1,
      success: true,
      message: None,
      body: Some(ResponseBody::Threads(ThreadsResponse::default())),
    };
    let value = resp_to_value(&resp);
    assert!(SCHEMA.is_valid(&value));
  }
```

So this relies on the `Default` impl for each type. Obviously, this will mean that any `Option`s will be `None` so many fields are not actually tested. But this is a good starting point. `syn` can be used to discover the struct fields and set them to generate more kinds of tests. 

DRAFT: currently I cannot make the tests fail by renaming a field completely so I'm doing something wrong.

### CODE REVIEW

File: integration_tests/Cargo.toml
Comment (pileghoff) [inline integration_tests/Cargo.toml:15]: Any reason for using syn version 1 instead of 2?

File: integration_tests/build.rs
Comment (pileghoff) [inline integration_tests/build.rs:36]: Couldn't you use something like to [quote](https://docs.rs/quote/latest/quote/index.html) to generate the source code, instead of using string formatting? That gives us syntax highlighting and makes things a bit more readable, in my opinion.

**Review by pileghoff — COMMENTED** [review]
Other than two small things, it looks good 👍🏽

**Comment by pileghoff** [comment]
Also, feel free to add me as a maintainer 😀

**Comment by sztomi** [comment]
Finally, some progress (and lots of broken responses, unsurprisingly):

image

**Comment by sztomi** [comment]
@pileghoff the generated tests are now passing, please give this another go if you don't mind. Most of the time the required fix was not serializing `None`/`null`. While I think that shouldn't necessarily break the schema validation, I'm fine with doing this because the generated JSON will be smaller. 

I will likely cut a release after this is merged. 

For a bit more sophisticated testing I will generate structs with random data (that's not null). I don't know if there's already some crate out there that does that but in the worst case I can write a macro for it. 

For testing the requests deserialization, a similar approach can be followed, but by generating JSON first and deserializing that.

**Comment by pileghoff** [comment]
I hope to look at it over the weekend 👍🏽 I was actually about to ask about making a new release with all the new fixes. I think after this is merged is a great time.

Regarding generating test data, would it be worth creating an example application that could talk to some of the most popular editors, so we could record the interaction and get some data that way? Then we could ensure that we were not only compliant with the spec, but that we also (sort of) works with those editors, in case they have some quirks or additions?

**Comment by sztomi** [comment]
> Regarding generating test data, would it be worth creating an example application that could talk to some of the most popular editors, so we could record the interaction and get some data that way?

Good idea and probably more useful than a synthetic test for the requests. They might do things slightly differently and dap-rs should work with all of them.

File: integration_tests/debugAdapterProtocol.json
Comment (pileghoff) [inline integration_tests/debugAdapterProtocol.json:1]: I think it would be nice to have some version information attached to this file somehow. Maybe we should add a readme file to the `integration_tests` folder, which includes the git repo and sha where this file was downloaded. That way to can keep track in case the spec is later updated.

File: integration_tests/src/lib.rs
Comment (pileghoff) [inline integration_tests/src/lib.rs]: I think we need a better comment here explaining what you are trying to do, as its not obivious. Just a couple of lines to make the reader decide if they want to dive in 😁 We could also link the [issue](https://github.com/Stranger6667/jsonschema-rs/issues/432) you posted, and which i stumbled over when trying to figure out what you were doing here 😆

File: integration_tests/build.rs
Comment (pileghoff) [inline integration_tests/build.rs:20]: Could we use the [cargo](https://doc.rust-lang.org/cargo/reference/config.html#env) to set an env variables, to replace some of the hardcoded filenames,such as the jsonschema file and `generated_tests.rs`?

File: integration_tests/src/lib.rs
Comment (pileghoff) [inline integration_tests/src/lib.rs]: Is this used anywhere?

**Review by pileghoff — COMMENTED** [review]
The tests are passing, which is the most important thing 🥇 I have a few cosmetic suggestion, but its great to see how many errors you were able to catch here, so we dont have to fix them individually as users encounters them.

File: integration_tests/src/lib.rs
Comment (sztomi) [reply integration_tests/src/lib.rs]: Fair enough, I added a longer comment and an example.

File: integration_tests/src/lib.rs
Comment (sztomi) [reply integration_tests/src/lib.rs]: Good catch, not anymore. And `lazy_static` isn't either, so one less dependency.

File: integration_tests/build.rs
Comment (sztomi) [reply integration_tests/build.rs:20]: We could but I'm not sure that's a customization point that solves a problem. Do you have a specific use-case in mind?

File: integration_tests/debugAdapterProtocol.json
Comment (sztomi) [reply integration_tests/debugAdapterProtocol.json:1]: Very good point. I changed the filename to reflect the commit in https://github.com/microsoft/debug-adapter-protocol/ and added a README explaining the convention.

File: integration_tests/build.rs
Comment (pileghoff) [reply integration_tests/build.rs:20]: Thinking about it, env variables is not the right place to put it. I just would like it better if we defined the file name somewhere and used that define instead of hard coding the path. The goal was not to make it customizable, using cargo was just the  only way i could see us doing it.

Its also just a minor nit 😁

Comment (sztomi) [reply integration_tests/build.rs:20]: It could be set in a config.toml and read via `env!`, but I'm not sure it's worth the complexity. `generated_sources.rs` is a very obvious name that communicates intent in both locations. Plus using an env variable just means replacing one string for another (the name of the environment variable).

**Review by pileghoff — APPROVED** [review]
I think it looks good 👍🏽

**Pull request merged**
